### PR TITLE
nuke tcp.Address and add proper IPv6 support

### LINKED ISSUE #1994: Cannot bind mitmproxy to IPv6 address
##### Steps to reproduce the problem:

1. `mitmproxy -b ::1`
`Error starting proxy server: gaierror(-9, 'Address family for hostname not supported')`

##### Any other comments? What have you tried so far?
The relevant code is probably hardcoded to IPv4  
https://github.com/mitmproxy/mitmproxy/blob/2316c0fb74efefd76970c685b8f77f45834f1490/mitmproxy/net/tcp.py#L921

Original Discussion:
https://discourse.mitmproxy.org/t/serving-on-ipv6/331

### PR DESCRIPTION
**Pull request by Kriechi**
This PR helps us to get IPv6-ready.

fixes #1994
closes #1995

### CODE REVIEW

**Comment by Kriechi** [comment]
@mhils: only 1 failing test, that is related to binding to an IPv6 socket. At the moment I hard-coded it to use the v4 family. We need to introduce a nifty way of fixing this. The second example could be useful: https://docs.python.org/3/library/socket.html#example

@ujjwal96 ping FYI

File: mitmproxy/flowfilter.py [hunk 1]
```diff
@@ -357,7 +360,10 @@ class FDst(_Rex):
     is_binary = False
 
     def __call__(self, f):
-        return f.server_conn.address and self.re.search(repr(f.server_conn.address))
```
Comment (mhils) [inline mitmproxy/flowfilter.py]: see https://github.com/mitmproxy/mitmproxy/pull/2050/files#r102108753

File: mitmproxy/net/tcp.py [hunk 2]
```diff
@@ -915,19 +821,18 @@ class TCPServer:
     request_queue_size = 20
 
     def __init__(self, address):
-        self.address = Address.wrap(address)
+        self.address = address
         self.__is_shut_down = threading.Event()
         self.__shutdown_request = False
-        self.socket = socket.socket(self.address.family, socket.SOCK_STREAM)
+        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (mhils) [inline mitmproxy/net/tcp.py]: As you mentioned, this is why IPv6 binding fails. We probably should implement the strategy proposed in https://github.com/mitmproxy/mitmproxy/pull/1995 here: Try IPv6 with V6ONLY=false, and fall back to IPv4 if that fails. We can do that in a separate PR if you want.

File: mitmproxy/io_compat.py [hunk 3]
```diff
@@ -88,6 +88,8 @@ def convert_019_100(data):
 
 def convert_100_200(data):
     data["version"] = (2, 0, 0)
+    data["client_conn"]["address"] = data["client_conn"]["address"]["address"]
+    data["server_conn"]["address"] = data["server_conn"]["address"]["address"]
```
Comment (Kriechi) [inline mitmproxy/io_compat.py]: Done.

File: mitmproxy/flowfilter.py (hunk 1 above)
Comment (Kriechi) [inline mitmproxy/flowfilter.py]: Done.

File: mitmproxy/flowfilter.py [hunk 4]
```diff
@@ -348,7 +348,10 @@ class FSrc(_Rex):
     is_binary = False
 
     def __call__(self, f):
-        return f.client_conn.address and self.re.search(repr(f.client_conn.address))
+        r = repr(f.client_conn.address)
```
Comment (Kriechi) [inline mitmproxy/flowfilter.py]: Done.

File: mitmproxy/net/tcp.py [hunk 5]
```diff
@@ -743,12 +653,12 @@ def convert_to_ssl(self, sni=None, alpn_protos=None, **sslctx_kwargs):
 
     def makesocket(self):
```
Comment (Kriechi) [inline mitmproxy/net/tcp.py]: Does the cuckoo sandbox team use this API? I would like to limit the API changes if possible.
Or do you know how they use it?

**Comment by mhils** [comment]
Copying here from Slack: We can't use `create_connection` as we need to do the `spoof_source_address` logic between `s = socket.socket()` and `s.connect()`. We need to so something like this:

1. Move `if self.spoof_source_address:` and `connection.bind(self.source_address)` into `_makesocket()`. 
2. Copy the implementation of `create_connection` and adapt it to use our _makesocket: https://hg.python.org/cpython/file/3.6/Lib/socket.py#l688

File: mitmproxy/net/tcp.py [hunk 6]
```diff
@@ -915,19 +852,36 @@ class TCPServer:
...
+        self.address = address
         self.__is_shut_down = threading.Event()
         self.__shutdown_request = False
-        self.socket = socket.socket(self.address.family, socket.SOCK_STREAM)
-        self.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-        self.socket.bind(self.address())
-        self.address = Address.wrap(self.socket.getsockname())
+
+        if self.address == 'localhost':
+            raise ValueError("Binding to 'localhost' is prohibited. Please use '::1' or '127.0.0.1' directly.")
```
Comment (mhils) [inline mitmproxy/net/tcp.py]: socket.error here so that this function only raises this?

**Comment by Kriechi** [comment]
Yes, I grep-ed for `*.host` and `*.port` in `console/`.

**Comment by Kriechi** [comment]
Done.

**Pull request merged**